Include/cpython/pyatomic.h

```diff
@@ -0,0 +1,397 @@
+
+static inline unsigned int
+_Py_atomic_load_uint_relaxed(const unsigned int *address);
+
+static inline Py_ssize_t
+_Py_atomic_load_ssize_relaxed(const Py_ssize_t *address);
+
+static inline void *
+_Py_atomic_load_ptr_relaxed(const void *address);
+
+// Atomically performs `*address = value` (sequential consistency)
+static inline void
+_Py_atomic_store_int(int *address, int value);
+
+static inline void
+_Py_atomic_store_int8(int8_t *address, int8_t value);
+
+static inline void
+_Py_atomic_store_int16(int16_t *address, int16_t value);
+
+static inline void
+_Py_atomic_store_int32(int32_t *address, int32_t value);
+
+static inline void
+_Py_atomic_store_int64(int64_t *address, int64_t value);
+
+static inline void
+_Py_atomic_store_intptr(intptr_t *address, intptr_t value);
+
+static inline void
+_Py_atomic_store_uint8(uint8_t *address, uint8_t value);
+
+static inline void
+_Py_atomic_store_uint16(uint16_t *address, uint16_t value);
+
+static inline void
+_Py_atomic_store_uint32(uint32_t *address, uint32_t value);
+
+static inline void
+_Py_atomic_store_uint64(uint64_t *address, uint64_t value);
+
+static inline void
+_Py_atomic_store_uintptr(uintptr_t *address, uintptr_t value);
+
+static inline void
+_Py_atomic_store_uint(unsigned int *address, unsigned int value);
+
+static inline void
+_Py_atomic_store_ptr(void *address, void *value);
+
+static inline void
+_Py_atomic_store_ssize(Py_ssize_t* address, Py_ssize_t value);
+
+// Stores `*address = value` (relaxed consistency, i.e., no ordering)
+static inline void
+_Py_atomic_store_int_relaxed(int *address, int value);
+
+static inline void
+_Py_atomic_store_int8_relaxed(int8_t *address, int8_t value);
+
+static inline void
+_Py_atomic_store_int16_relaxed(int16_t *address, int16_t value);
+
+static inline void
+_Py_atomic_store_int32_relaxed(int32_t *address, int32_t value);
+
+static inline void
+_Py_atomic_store_int64_relaxed(int64_t *address, int64_t value);
+
+static inline void
+_Py_atomic_store_intptr_relaxed(intptr_t *address, intptr_t value);
+
+static inline void
+_Py_atomic_store_uint8_relaxed(uint8_t* address, uint8_t value);
+
+static inline void
+_Py_atomic_store_uint16_relaxed(uint16_t *address, uint16_t value);
+
+static inline void
+_Py_atomic_store_uint32_relaxed(uint32_t *address, uint32_t value);
+
+static inline void
+_Py_atomic_store_uint64_relaxed(uint64_t *address, uint64_t value);
+
+static inline void
+_Py_atomic_store_uintptr_relaxed(uintptr_t *address, uintptr_t value);
+
+static inline void
+_Py_atomic_store_uint_relaxed(unsigned int *address, unsigned int value);
+
+static inline void
+_Py_atomic_store_ptr_relaxed(void *address, void *value);
+
+static inline void
+_Py_atomic_store_ssize_relaxed(Py_ssize_t *address, Py_ssize_t value);
+
+// Stores `*address = value` (release operation)
+static inline void
+_Py_atomic_store_uint64_release(uint64_t *address, uint64_t value);
+
+static inline void
+_Py_atomic_store_ptr_release(void *address, void *value);
```

Comment:
I'm not an expert, but why is it useful to expose "release" operations if no "acquire" operations are exposed?

Reply:
You can always use stronger orderings for correctness (i.e., "seq_cst" everywhere instead of "acquire"). From a performance view, "release" is substantially faster than "seq_cst" stores on x86/x86-64, but "acquire" generates the same code as "seq_cst" loads on both x86/x86-64 and aarch64.

Reply:
I don't think CPython support is limited to x86 and ARM variants, so the set of atomic ops exposed should probably be made consistent nevertheless?

Also, using "seq_cst" in combination with "release" will probably make the code more difficult to reason about, than if "acquire" is exposed (and memory ordering is already hard to reason about!).